group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntimeTest.java

```diff
@@ -0,0 +1,803 @@
+        assertEquals(Arrays.asList(0L, 2L), ctx.snapshotRegistry.epochsList());
+        // Records have been replayed to the coordinator.
+        assertEquals(mkSet("record1", "record2"), ctx.coordinator.records());
+        // Records have been written to the log.
+        assertEquals(Arrays.asList("record1", "record2"), writer.records(TP));
+
+        // Write #2.
+        CompletableFuture<String> write2 = runtime.scheduleWriteOperation("write#2", TP,
+            state -> new CoordinatorResult<>(Arrays.asList("record3"), "response2"));
+
+        // Verify that the write is not completed yet.
+        assertFalse(write2.isDone());
+
+        // The last written offset is updated.
+        assertEquals(3L, ctx.lastWrittenOffset);
+        // The last committed offset does not change.
+        assertEquals(0L, ctx.lastCommittedOffset);
+        // A new snapshot is created.
+        assertEquals(Arrays.asList(0L, 2L, 3L), ctx.snapshotRegistry.epochsList());
+        // Records have been replayed to the coordinator.
+        assertEquals(mkSet("record1", "record2", "record3"), ctx.coordinator.records());
+        // Records have been written to the log.
+        assertEquals(Arrays.asList("record1", "record2", "record3"), writer.records(TP));
+
+        // Write #3 but without any records.
+        CompletableFuture<String> write3 = runtime.scheduleWriteOperation("write#3", TP,
+            state -> new CoordinatorResult<>(Collections.emptyList(), "response3"));
+
+        // Verify that the write is not completed yet.
+        assertFalse(write3.isDone());
+
+        // The state does not change.
+        assertEquals(3L, ctx.lastWrittenOffset);
+        assertEquals(0L, ctx.lastCommittedOffset);
+        assertEquals(Arrays.asList(0L, 2L, 3L), ctx.snapshotRegistry.epochsList());
+        assertEquals(mkSet("record1", "record2", "record3"), ctx.coordinator.records());
+        assertEquals(Arrays.asList("record1", "record2", "record3"), writer.records(TP));
+
+        // Commit write #1.
+        writer.commit(TP, 2);
+
+        // The write is completed.
+        assertTrue(write1.isDone());
+        assertEquals("response1", write1.get(5, TimeUnit.SECONDS));
+
+        // The last committed offset is updated.
+        assertEquals(2L, ctx.lastCommittedOffset);
+        // The snapshot is cleaned up.
+        assertEquals(Arrays.asList(2L, 3L), ctx.snapshotRegistry.epochsList());
+
+        // Commit write #2.
+        writer.commit(TP, 3);
+
+        // The writes are completed.
+        assertTrue(write2.isDone());
+        assertTrue(write3.isDone());
+        assertEquals("response2", write2.get(5, TimeUnit.SECONDS));
+        assertEquals("response3", write3.get(5, TimeUnit.SECONDS));
+
+        // The last committed offset is updated.
+        assertEquals(3L, ctx.lastCommittedOffset);
+        // The snapshot is cleaned up.
+        assertEquals(Collections.singletonList(3L), ctx.snapshotRegistry.epochsList());
```

Comment:
just for my understanding, we will always keep the latest snapshot in case we need to rollback?